### ISSUE TITLE
fileio.c: verify_file broken for sysfs writes on s390

### ISSUE THREAD

**Issue by tblume** [systemd/systemd#19314]
**systemd version the issue has been seen with**
 >246 

**Used distribution**
 > SUSE Linux Enterprise Server 15 SP3

**Linux kernel version used** (`uname -a`)

 > ` Linux s390vsl004 5.3.18-53-default #1 SMP Wed Mar 31 13:11:52 UTC 2021 (85a0bf4) s390x s390x s390x GNU/Linux`

**CPU architecture issue was seen on**
 > s390x

**Expected behaviour you didn't see**
 > no error message on multiple writes into sysfs files, for example:
> /sys/devices/qeth/0.0.0800/portno

**Unexpected behaviour you saw**
 >Apr 13 16:10:33 s390vsl004 systemd-udevd[807]: 0.0.0800: /etc/udev/rules.d/41-qeth-0.0.0800.rules:18 Failed to write ATTR{/sys/devices/qeth/0.0.0800/portno}, ignoring: Operation not permitted
>Apr 13 16:10:33 s390vsl004 systemd-udevd[807]: 0.0.0800: /etc/udev/rules.d/41-qeth-0.0.0800.rules:19 Failed to write ATTR{/sys/devices/qeth/0.0.0800/layer2}, ignoring: Operation not permitted
>Apr 13 16:10:33 s390vsl004 systemd-udevd[807]: 0.0.0800: /etc/udev/rules.d/41-qeth-0.0.0800.rules:20 Failed to write ATTR{/sys/devices/qeth/0.0.0800/online}, ignoring: Invalid argument

Actually it is normal that writes to those sysfs files only succeed first time and subsequent writes fail.
verify_file should be able to cope with that since the same value is written again. 

**Steps to reproduce the problem**
 > journalctl -axb | grep 'Failed to write'

 > udevadm test /sys/devices/qeth/0.0.0800
 > (requires udev rules generated by chzdev which is done during installation)

**Additional program output to the terminal or log subsystem illustrating the issue**

with the following instrumentation:

```
diff --git a/src/basic/fileio.c b/src/basic/fileio.c
index c3d55d209a..690a3ae40e 100644
--- a/src/basic/fileio.c
+++ b/src/basic/fileio.c
@@ -284,8 +284,10 @@ fail:
          * contents in place already. If so, eat up the error. */
 
         q = verify_file(fn, line, !(flags & WRITE_STRING_FILE_AVOID_NEWLINE));
-        if (q <= 0)
+        if (q <= 0) {
+                log_warning("tblume: verify_file failed for %s retval is %d returning %d", fn, q, r);
                 return r;
+        }
 
         return 0;
 }
@@ -351,8 +353,10 @@ int verify_file(const char *fn, const char *blob, bool accept_extra_nl) {
         if (ferror(f))
                 return errno_or_else(EIO);
 
-        if (k != l && k != l + accept_extra_nl)
+        if (k != l && k != l + accept_extra_nl) {
+                log_warning("tblume: k !=l succeeded, k is %ld, l is %ld l + newline is %ld", k, l, (l + accept_extra_nl));
                 return 0;
+        }
         if (memcmp(buf, blob, l) != 0)
                 return 0;
         if (k > l && buf[l] != '\n')
```

I get the udevadm test output:

```
path '[ccwgroup/0.0.0800]portno' is '/sys/devices/qeth/0.0.0800/portno'
0.0.0800: /etc/udev/rules.d/41-qeth-0.0.0800.rules:18 ATTR '/sys/devices/qeth/0.0.0800/portno' writing '0'
tblume: k !=l succeeded, k is 2, l is 1 l + newline is 1
tblume: verify_file failed for /sys/devices/qeth/0.0.0800/portno retval is 0 returning -1
0.0.0800: /etc/udev/rules.d/41-qeth-0.0.0800.rules:18 Failed to write ATTR{/sys/devices/qeth/0.0.0800/portno}, ignoring: Operation not permitted
```

**Comment by yuwata** [comment]
> Actually it is normal that writes to those sysfs files only succeed first time and subsequent writes fail.
verify_file should be able to cope with that since the same value is written again.

I do not catch this. `verify_file()` does not write anything, but read the file.

It seems just kernel simply refuses to write the files. Are these files really writable?

**Comment by tblume** [comment]
> > Actually it is normal that writes to those sysfs files only succeed first time and subsequent writes fail.
> > verify_file should be able to cope with that since the same value is written again.
> 
> I do not catch this. `verify_file()` does not write anything, but read the file.
> 
> It seems just kernel simply refuses to write the files. Are these files really writable?

Yes those files are writable:

`s390vsl004:~ # ls -l /sys/devices/qeth/0.0.0800/portno
-rw-r--r-- 1 root root 4096 14. Apr 15:12 /sys/devices/qeth/0.0.0800/portno`

but only the first time. A subsequent write, even with the same value fails.
According to the code verify_file should handle this:

`       r = write_string_stream_ts(f, line, flags, ts);
        if (r < 0)
                goto fail;

        return 0;

fail:
        if (!(flags & WRITE_STRING_FILE_VERIFY_ON_FAILURE))
                return r;

        f = safe_fclose(f);

        /* OK, the operation failed, but let's see if the right
         * contents in place already. If so, eat up the error. */

        q = verify_file(fn, line, !(flags & WRITE_STRING_FILE_AVOID_NEWLINE));`

**Comment by yuwata** [comment]
What did you write at the first time? And, then at the second time? What is the current value?

**Comment by tblume** [comment]
> What did you write at the first time? And, then at the second time? What is the current value?

The write is done by udev rules:
```
# cat /etc/udev/rules.d/41-qeth-0.0.0800.rules
# Generated by chzdev
ACTION=="add", SUBSYSTEM=="drivers", KERNEL=="qeth", GOTO="group_qeth_0.0.0800"
ACTION=="add", SUBSYSTEM=="ccw", KERNEL=="0.0.0800", DRIVER=="qeth", GOTO="group_qeth_0.0.0800"
ACTION=="add", SUBSYSTEM=="ccw", KERNEL=="0.0.0801", DRIVER=="qeth", GOTO="group_qeth_0.0.0800"
ACTION=="add", SUBSYSTEM=="ccw", KERNEL=="0.0.0802", DRIVER=="qeth", GOTO="group_qeth_0.0.0800"
ACTION=="add", SUBSYSTEM=="ccwgroup", KERNEL=="0.0.0800", DRIVER=="qeth", GOTO="cfg_qeth_0.0.0800"
GOTO="end_qeth_0.0.0800"

LABEL="group_qeth_0.0.0800"
TEST=="[ccwgroup/0.0.0800]", GOTO="end_qeth_0.0.0800"
TEST!="[ccw/0.0.0800]", GOTO="end_qeth_0.0.0800"
TEST!="[ccw/0.0.0801]", GOTO="end_qeth_0.0.0800"
TEST!="[ccw/0.0.0802]", GOTO="end_qeth_0.0.0800"
ATTR{[drivers/ccwgroup:qeth]group}="0.0.0800,0.0.0801,0.0.0802"
GOTO="end_qeth_0.0.0800"

LABEL="cfg_qeth_0.0.0800"
ATTR{[ccwgroup/0.0.0800]portno}="0"
ATTR{[ccwgroup/0.0.0800]layer2}="1"
ATTR{[ccwgroup/0.0.0800]online}="1"

LABEL="end_qeth_0.0.0800"
```

For portno it is 0 for all writes and also the current value:

```
# cat /sys/devices/qeth/0.0.0800/portno
0
```

**Comment by tblume** [comment]
btw. it seems that before your commit 25de7aa7b90c23d33ea50ada1e50c5834a414237 that was completely broken, because fprintf was apparently doing buffered writes and the buffer happily accepted all writes, no matter whether the sysfs file would accept it on flush or not.
Therefore no such error was visible with older systemd versions.

**Comment by yuwata** [comment]
Hmm, so, kernel automatically adds `\n' at the end when reading `portno`??

**Comment by yuwata** [comment]
Hm, `qeth_dev_portno_show()` in drivers/s390/net/qeth_core_sys.c appends `\n`.

**Comment by tblume** [comment]
Sorry, I have no insight into IBM s390 specific kernel driver stuff. 
If you consider this rather a kernel bug, I would hand it over to IBM.

**Comment by yuwata** [comment]
Fix is waiting in #19318. But not tested, as I do not have such devices. If possible, could you test the PR?

**Comment by tblume** [comment]
> Fix is waiting in #19318. But not tested, as I do not have such devices. If possible, could you test the PR?

Will test ASAP and report.
Thanks for the fast response!

**Comment by tblume** [comment]
The test of the patch was successfull:
```
# udevadm test /sys/devices/qeth/0.0.0800
SELinux enabled state cached to: disabled
This program is for debugging only, it does not run any program
specified by a RUN key. It may show incorrect results, because
some values may be different, or not available at a simulation run.

Trying to open "/etc/systemd/hwdb/hwdb.bin"...
Trying to open "/etc/udev/hwdb.bin"...
=== trie on-disk ===
tool version:          246
file size:         9001605 bytes
header size             80 bytes
strings            2279613 bytes
nodes              6721912 bytes
Load module index
Found container virtualization none.
Network interface NamePolicy= disabled by default.
timestamp of '/etc/systemd/network' changed
timestamp of '/usr/lib/systemd/network' changed
Parsed configuration file /usr/lib/systemd/network/99-default.link
Created link configuration context.
timestamp of '/etc/udev/rules.d' changed
timestamp of '/run/udev/rules.d' changed
Reading rules file: /usr/lib/udev/rules.d/00-dont-del-part-nodes.rules
Reading rules file: /usr/lib/udev/rules.d/01-md-raid-creating.rules
[...]
Reading rules file: /usr/lib/udev/rules.d/99-systemd.rules
Reading rules file: /usr/lib/udev/rules.d/99-wakeup-from-idle.rules
path '[ccwgroup/0.0.0800]portno' is '/sys/devices/qeth/0.0.0800/portno'
0.0.0800: /etc/udev/rules.d/41-qeth-0.0.0800.rules:18 ATTR '/sys/devices/qeth/0.0.0800/portno' writing '0'
path '[ccwgroup/0.0.0800]layer2' is '/sys/devices/qeth/0.0.0800/layer2'
0.0.0800: /etc/udev/rules.d/41-qeth-0.0.0800.rules:19 ATTR '/sys/devices/qeth/0.0.0800/layer2' writing '1'
path '[ccwgroup/0.0.0800]online' is '/sys/devices/qeth/0.0.0800/online'
0.0.0800: /etc/udev/rules.d/41-qeth-0.0.0800.rules:20 ATTR '/sys/devices/qeth/0.0.0800/online' writing '1'
DEVPATH=/devices/qeth/0.0.0800
DEVTYPE=qeth_generic
DRIVER=qeth
ACTION=add
SUBSYSTEM=ccwgroup
USEC_INITIALIZED=352710554
Unload module index
Unloaded link configuration context.
```
